shark-hprof/src/main/java/shark/HprofDeobfuscator.kt

```diff
@@ -0,0 +1,237 @@
+      }
+
+    return maxId
+  }
+
+  @Suppress("LongParameterList")
+  private fun writeHprofRecords(
+    inputHprofFile: File,
+    outputHprofFile: File,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>,
+    classNames: MutableMap<Long, Long>,
+    firstId: Long
+  ): File {
+    var id = firstId
+
+    Hprof.open(inputHprofFile)
+      .use { hprof ->
+        val reader = hprof.reader
+        HprofWriter.open(
+          outputHprofFile,
+          identifierByteSize = reader.identifierByteSize,
+          hprofVersion = hprof.hprofVersion
+        ).use { writer ->
+          reader.readHprofRecords(setOf(HprofRecord::class),
+            OnHprofRecordListener { _,
+              record ->
+              // HprofWriter automatically emits HeapDumpEndRecord, because it flushes
+              // all continuous heap dump sub records as one heap dump record.
+              if (record is HeapDumpEndRecord) {
+                return@OnHprofRecordListener
+              }
+
+              when (record) {
+                is StringRecord -> {
+                  writer.write(
+                    createDeobfuscatedStringRecord(record, proguardMapping, hprofStringCache)
+                  )
+                }
+                is HeapDumpRecord -> {
+                  when (record) {
+                    is ObjectRecord -> {
+                      when (record) {
+                        is ClassDumpRecord -> {
+                          val recordsToWrite = createDeobfuscatedClassDumpRecord(
+                            record, proguardMapping, hprofStringCache, classNames, id
+                          )
+                          id += recordsToWrite.size
+                          recordsToWrite.forEach {
+                            writer.write(it)
+                          }
+                        }
+                        else -> writer.write(record)
+                      }
+                    }
+                    else -> writer.write(record)
+                  }
+                }
+                else -> writer.write(record)
+              }
+            })
+            }
+      }
+
+    return outputHprofFile
+  }
+
+  private fun createDeobfuscatedStringRecord(
+    record: StringRecord,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>
+  ): StringRecord {
+    val obfuscatedName = hprofStringCache[record.id]!!
+    return StringRecord(
+      record.id, proguardMapping.deobfuscateClassName(obfuscatedName)
+    )
+  }
+
+  /**
+   * Deobfuscated ClassDumpRecord's field names. Different classes can have fields with the same
+   * names. We need to generate new StringRecords in such cases.
+   *
+   * @return a list of HprofRecords to write.
+   */
+  private fun createDeobfuscatedClassDumpRecord(
+    record: ClassDumpRecord,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>,
+    classNames: MutableMap<Long, Long>,
```

Comment:
Why are we passing around mutable maps here? I'm not seeing where we update them.

Reply:
Right, I copy pasted it from `readHprofRecords` method. Will change it :)

Reply:
Done